### DISCUSSION TITLE
Relations within composite types

### DISCUSSION THREAD

**Discussion by jpteixeira99** [prisma/prisma#20101] [Q&A] [answered]
Given a schema such as:
```prisma
model User {
  id   String @id @map("_id")
  shop Shop
  ...
}

model ItemCategory {
  id         String                 @id @default(auto()) @map("_id") @db.ObjectId
  name       String
  ...
}

type Shop {
  items Item[]
  ...
}

type Item {
  name     String
  category   ItemCategory      @relation(fields: [categoryId], references: [id])
  categoryId String            @map("category") @db.ObjectId
  ...
}
```

Prisma complains that `Defining "@relation" attribute for a field in a composite type is not allowed`. What would be an alternative way of achieving the same result, if there is any?

**Comment by nurul3101** [comment, marked as answer]
Hello @Viriatto 👋 

One alternative way to achieve the same result is to transform your `Shop` and `Item` types into models and establish relationships between them and also between `Item` and `ItemCategory`.

Your schema could look something like this:

```prisma
model User {
  id String @id @default(auto()) @map("_id") @db.ObjectId
}

model ItemCategory {
  id    String @id @default(auto()) @map("_id") @db.ObjectId
  name  String
  items Item[] // The reverse relationship from ItemCategory to Item
}

model Shop {
  id    String @id @default(auto()) @map("_id") @db.ObjectId
  items Item[]
}

model Item {
  id         String       @id @default(auto()) @map("_id") @db.ObjectId 
  name       String
  category   ItemCategory @relation(fields: [categoryId], references: [id])
  categoryId String       @map("category") @db.ObjectId
  shop       Shop         @relation(fields: [shopId], references: [id]) 
  shopId     String       @map("shop") @db.ObjectId
}
```

It would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 
Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by jpteixeira99** [reply]
I guess my only concern with this approach is the fact it doesn't follow MongoDB's data modeling's main guideline which is to de-normalize data whenever possible... But it appears there isn't really a way to achieve a denormalized data model _and_ use relations within nested documents so I will have to pick one.

**Reply by nurul3101** [reply]
At the moment relations are only allowed on models and not on composite types, so you would need to use this approach.

**Reply by nurul3101** [reply]
Please feel free to let us know in case you have any additional queries.

If not, It would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 
Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by nurul3101** [reply]
We are closing this Discussion due to inactivity. Please open a new Discussion if you have a similar problem 🙏
